### ISSUE TITLE
Type error when using `useResource`

### ISSUE THREAD

**Issue by nightire** [NullVoxPopuli/ember-resources#48]
![image](https://user-images.githubusercontent.com/1104018/123830586-01683500-d936-11eb-89a1-e80ee853170a.png)

First of all, I can not write a failing test because in test cases, we instantiate the resource class directly, but the type error is caused by `useResource` utility function.

Then I spend a little time investigating what's going on with `useResource`, here when describing the arguments of `Constructable`, I found:

```typescript
export interface Constructable<T = unknown> {
    new (...args: unknown[]): T; // <- `unknown` causes the type error
}
```

If I change it to `any[]` or more precisely `(Positional | Named)[]`, the problem will be solved, at least for my case. I'm not sure if this fix is correct, @NullVoxPopuli how do you think?

**Comment by NullVoxPopuli** [comment]
Ah! That could be it. Yeah, I'd be happy with any there. It more or less is the same in this case anyway

**Comment by NullVoxPopuli** [comment]
does `Positional<[Field]>` work?

**Comment by nightire** [comment]
> does `Positional<[Field[]]>` work?

Unfortunately `Positional<[Field[]]>` still causes the same type error.

**Comment by NullVoxPopuli** [comment]
I made this demo, of what I thought was the issue, and the type checking passes: https://github.com/NullVoxPopuli/ember-resources/pull/49

Does this help?

**Comment by nightire** [comment]
> I made this demo, of what I thought was the issue, and the type checking passes: #49
> 
> Does this help?

Nope.

I'm starting to think maybe it's my side's problem, maybe I need to try in a new ember app instead.

**Comment by NullVoxPopuli** [comment]
Lemme know how it goes!

**Comment by nightire** [comment]
![image](https://user-images.githubusercontent.com/1104018/123986061-8f0c5900-d9f8-11eb-83b2-8f746354f6b1.png)

@NullVoxPopuli I can confirm the types are correct in a new ember app, now I need to figure out the problem of my side. Thank you anyway.

**Comment by nightire** [comment]
I reopen this issue because I found the root cause, I got this error before because I set `"strict": true` in `tsconfig.json`, `ember-cli-typescript` doesn't use this flag by default, but out company follows a more strict policy.

Since `"strict": true` means to enabled a wide range of type checking, I try to toggle them one by one and I'm pretty sure `"strictFunctionTypes": true` will cause this error. According to its [documentation](https://www.typescriptlang.org/tsconfig#strictFunctionTypes), I think it does make sense:

> When enabled, this flag causes functions parameters to be checked more correctly.

**Comment by NullVoxPopuli** [comment]
@nightire do you happen to have a reproduction repo?

**Comment by nightire** [comment]
Sorry for the delay, I've got caught up in something else this last month.

Here is a reproduction: https://github.com/nightire/ember-resources-type-issue/tree/main#ember-resources-type-issue, hope it can help you.

**Comment by NullVoxPopuli** [comment]
@nightire can you try this latest release?, 3.2.1?

**Comment by nightire** [comment]
> @nightire can you try this latest release?, 3.2.1?

Unfortunately, the error remains. I can live with `"strictFunctionTypes": false` for now.

**Comment by NullVoxPopuli** [comment]
Does the error in your reproduction go away?
What's the reproduction missing? 🤔

**Comment by nightire** [comment]
> Does the error in your reproduction go away?
> What's the reproduction missing? 🤔

I have upgraded the ember-resources in my reproduction, and the error remains. 

This repro is a brand new v3.28.0 ember application with the latest ember-cli-typescript and ember-resources, nothing else.

BTW, this small change will solve the error as I mentioned at the beginning of this issue. Is it an acceptable solution? I'm not an expert on typescript, just ask for curiosity.

```diff
export interface Constructable<T = unknown> {
-  new (...args: unknown[]): T;
+  new (...args: any[]): T;
}
```

**Comment by lougreenwood** [comment]
I also just hit this issue after enabling `strict` TS mode.

**Comment by nightire** [comment]
> I also just hit this issue after enabling `strict` TS mode.

You can use `strict` mode with `strictFunctionType` off:

```
  "strict": true,
  "strictFunctionTypes": false,
```

I found this setting is not that important as we thought it would be, according to the official document:

> During the development of this feature, we discovered a large number of inherently unsafe class hierarchies, including some in the DOM. Because of this, the setting only applies to functions written in function syntax, not to those in method syntax

**Comment by NullVoxPopuli** [comment]
fixed via... deprecation :sweat_smile: 

better alternative landed in v4.5
